lib/elixir/test/elixir/module/types/descr_test.exs

```diff
@@ -1298,17 +1318,50 @@ defmodule Module.Types.DescrTest do
              |> to_quoted_string() ==
                "non_empty_list(term()) and not (non_empty_list(atom()) or non_empty_list(integer()))"
 
+      assert list(integer()) |> union(list(atom())) |> to_quoted_string() ==
+               "list(atom()) or list(integer())"
+
+      assert list(integer())
+             |> union(list(atom()))
+             |> difference(empty_list())
+             |> to_quoted_string() ==
+               "non_empty_list(atom()) or non_empty_list(integer())"
+
       assert list(term(), integer()) |> to_quoted_string() ==
-               "empty_list() or non_empty_list(term(), integer())"
+               "maybe_improper_list(term(), integer())"
 
       assert difference(list(term(), atom()), list(term(), boolean())) |> to_quoted_string() ==
-               "non_empty_list(term(), atom() and not boolean())"
+               "non_empty_maybe_improper_list(term(), atom() and not boolean())"
 
       assert list(term(), term()) |> to_quoted_string() ==
-               "empty_list() or non_empty_list(term(), term())"
+               "maybe_improper_list(term(), term())"
 
-      assert non_empty_list(term(), difference(term(), list(term()))) |> to_quoted_string() ==
-               "improper_list()"
+      assert non_empty_maybe_improper_list(term(), term()) |> to_quoted_string() ==
+               "non_empty_maybe_improper_list(term(), term())"
+
+      # TODO this feels wrong, figure this out
+      # seems to be due to how list_descr discards negations, not an issue with to_quoted_string
```

Comment:
I believe this is correct. Because `difference(term(), empty_list())` still includes `maybe_improper_list(term(), term())`, that list may still end with an `empty_list()`, and therefore this list can still finish with an empty list.

Think about it like this:

```
certainly_not_empty_tail = [1, 2, 3 | [4, 5, 6 | what_is_this]]
```

Depending on `what_is_this`, `certainly_not_empty_tail` will end with an empty list as tail.

Reply:
🤯 got it !